### DISCUSSION TITLE
Feedback on Using Medplum for FHIR Crud Operations

### DISCUSSION THREAD

**Discussion by khanuzair** [medplum/medplum#4658] [General]
Hi folks, big fan of Medplum! I've been doing a deep dive of the codebase this week to understand it a bit better, really like how organized the codebase is throughout. At this point I'm not sure if I can use Medplum for the use case I had in mind so hoping to get some feedback; esp from @codyebberson and others on the team. 

The use case I'm looking for is to use a common library that can handle CRUD operations for FHIR resources while handling the low level Postgres sql construction/querying/parsing. And being able to connect to my Postgres db in Supabase.

I'm running the Medplum server locally and changed the `medplum.config.json` to with the creds in Supabase which does work. The issue is I don't want to stand up a FHIR server with login, etc since the service I'm building will already handle those parts. I really just want an easy way to use the `MedplumClient` to do operations on FHIR resources and trust that it will get populated in the database and is returned back in a parsed way. 

I'm not really seeing a good way of doing this at the moment; is the recommended path here to run to run the FHIR server ourselves with a single (almost dummy) service login creds? One option I've thought about is using a cloud FHIR datastore, like what Google Cloud offers, but I'd really like to use Medplum since it's open source and leaves no ambiguity about how it works end to end.

**Comment by khanuzair** [comment]
Or another follow up, is it possible to use the `MedplumClient` with an external FHIR server like Google Cloud Healthcare API or AWS Healthlake? It is unclear if this is possible from the docs, I do see an example with Epic but don't see in the docs whether it's possible to do it with one of the cloud providers or not.

**Comment by codyebberson** [comment]
Hi @khanuzair - Thanks for your kind words.  Apologies for the slow reply, I was out on holiday.

I think I understand your desired architecture.  Unfortunately, we don't have exactly what you're looking for at the moment.

Today, yes, the answer is what you suggested: "run to run the FHIR server ourselves with a single (almost dummy) service login creds".

We're actively building out that use case, with features such as "On-Behalf-Of" (https://github.com/medplum/medplum/issues/3892) so that you can still take advantage of user-based permissions and whatnot.

Avoiding a separate server is not possible at the moment.  A while back, we contemplated breaking up `server` into smaller components which could be referenced as Express middleware, which would basically allow you to run Medplum "server" inside your own server app.  See old discussion here: https://github.com/medplum/medplum/discussions/1424  

> is it possible to use the MedplumClient with an external FHIR server like Google Cloud Healthcare API or AWS Healthlake?

Yes, we frequently use `MedplumClient` with other FHIR servers.  I have not tried it with GCP or AWS Healthlake, so I can't speak directly to those two.  The most common challenge is non-standard auth, which can require extra work to figure out custom HTTP headers and whatnot.  Most methods on `MedplumClient` take an optional `MedplumRequestOptions` param gets passed to `fetch`

Please let me know if you have any additional questions.  We'd be happy to get creative to fit your desired architecture.
